Sources/SPMBuildCore/ExtensionEvaluator.swift

```diff
@@ -0,0 +1,328 @@
+        }
+        return (output: output, stdoutText: stdoutText)
+    }
+}
+
+
+/// Represents the result of evaluating an extension against a particular resolved-target.  This includes generated
+/// commands as well as any diagnostics or output emitted by the extension.
+public struct ExtensionEvaluationResult {
+    /// The extension that produced the results.
+    public let `extension`: ExtensionTarget
+    
+    /// The commands generated by the extension (in order).
+    public let commands: [Command]
+
+    /// A command provided by an extension. Extensions are evaluated after package graph resolution (and subsequently,
+    /// if conditions change). Each extension specifies capabilities the capability it provides, which determines what
+    /// kinds of commands it generates (when they run during the build, and the specific semantics surrounding them).
+    public enum Command {
+        
+        /// A command to run before the start of every build. Besides the obvious parameters, it can provide a list of
+        /// directories whose contents should be considered as inputs to the set of source files to which build rules
+        /// should be applied.
+        case prebuildCommand(
+                displayName: String,
+                execPath: AbsolutePath,
+                arguments: [String],
+                workingDir: AbsolutePath?,
+                environment: [String: String]?,
+                derivedSourceDirPaths: [AbsolutePath]
+             )
+        
+        /// A command to be incorporated into the build graph, so that it runs when any of the outputs are missing or
+        /// the inputs have changed from the last time when it ran. This is the preferred kind of command to generate
+        /// when the input and output paths are known. In addition to inputs and outputs, the command can specify one
+        /// or more files that should be considered as inputs to the set of source files to which build rules should
+        /// be applied.
+        case buildToolCommand(
+                displayName: String,
+                execPath: AbsolutePath,
+                arguments: [String],
+                workingDir: AbsolutePath?,
+                environment: [String: String]?,
+                inputPaths: [AbsolutePath],
+                outputPaths: [AbsolutePath],
+                derivedSourcePaths: [AbsolutePath]
+             )
+        
+        /// A command to run after the end of every build.
+        case postbuildCommand(
+                displayName: String,
+                execPath: AbsolutePath,
+                arguments: [String],
+                workingDir: AbsolutePath?,
+                environment: [String: String]?
+             )
+    }
+    
+    // Any diagnostics emitted by the extension.
+    public let diagnostics: [Diagnostic]
+    
+    // A location representing a file name or path and an optional line number.
+    // FIXME: This should be part of the Diagnostics APIs.
```

Comment:
maybe add as an extension to Basics until we move to TSC

Reply:
That makes sense.  I think maybe that could be a separate follow-on PR?